cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -81,91 +81,109 @@ def process_data(old_data, new_data)
+        values = old_vars.merge(values)
       end
+      set_env_variables(stackname, current_grid) # set envs for execution time
+      parsed_stack = reader.execute(
+        values: values,
+        defaults: old_data[stackname].nil? ? nil : old_data[stackname][:stack_data]['variables'],
+        parent_name: data[:parent_name],
+        name: data[:name]
+      )
+      parsed_stack["parent"] = { "name" => parsed_stack.delete("parent_name") }
+      data[:stack_data] = parsed_stack
+    ensure
+      prev_env.each { |k, v| ENV[k] = v }
+    end
 
-      puts "STACKS:"
-      puts "-" * 40
+    # @param changes [Kontena::Stacks::ChangeResolver]
+    def display_report(changes)
+      puts
+      caret "Calculated changes:", dots: false
 
       unless changes.removed_stacks.empty?
-        puts pastel.red("These stacks #{will} be removed because they are no longer depended on:")
-        changes.removed_stacks.each { |stack| puts pastel.red("- #{stack}") }
-        puts
+        changes.removed_stacks.each { |stack|
+          puts "  #{pastel.red('-')} #{stack}"
+          changes.removed_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.red('-')} #{s}"
+          end
+        }
       end
 
       unless changes.replaced_stacks.empty?
-        puts pastel.yellow("These stacks #{will} be replaced by other stacks:")
+        changes.replaced_stacks.each { |stack|
+          puts "  #{pastel.yellow('-/+')} #{stack}"
+          changes.upgraded_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.cyan('~')} #{s}"
+          end
+          changes.added_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.green('+')} #{s}"
+          end
+          changes.removed_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.red('-')} #{s}"
+          end
+        }
         changes.replaced_stacks.each do |installed_name, data|
           puts "- #{pastel.yellow(installed_name)} from #{pastel.cyan(data[:from])} to #{pastel.cyan(data[:to])}"
         end
         puts
       end
 
       unless changes.added_stacks.empty?
-        puts pastel.green("These new stack dependencies #{will} be installed:")
-        changes.added_stacks.each { |stack| puts pastel.green("- #{stack}") }
-        puts
+        changes.added_stacks.each { |stack|
+          puts "  #{pastel.green('+')} #{stack}"
+          changes.added_services.select { |s| s.split('/')[0] == stack }.each do |s|
+            puts "    #{pastel.green('+')} #{s}"
+          end
+        }
       end
 
       unless changes.upgraded_stacks.empty?
-        puts pastel.cyan("These stacks #{will} be upgraded#{' and deployed' if deploy?}:")
-        changes.upgraded_stacks.each { |stack| puts pastel.cyan("- #{stack}") }
-        puts
+        changes.upgraded_stacks.each { |stack|
```

Comment:
Could this also explicitly list the dependent stacks that are *NOT* getting upgraded?

Right now it looks a little broken to me, because it will only list the top-level stack:

```
 [done] Parsing test/deptest.yaml     
DEBUG Master stacks: deptest,deptest-bar,deptest-foo YAML stacks: deptest,deptest-foo,deptest-bar
 [done] Processing stack deptest-bar
 [done] Processing stack deptest-foo
Value for peers is not set. Substituting with an empty string.
 [done] Processing stack deptest
 [done] Analyzing upgrade     

> Calculated changes:
  ~ deptest
    ~ deptest/redis

    *   Upgrading stack deptest .. |
```

Reply:
It could, was not really sure does that make sense? Output show "changes"...

Reply:
It's a little hard to tell exactly what stacks/services you have present in a complex set of dependent stacks. I'd be more comfortable seeing the complete list of stacks including an explicit list of things not changed, rather than having to wonder if I somehow accidentally mis-indented the YAML and all the dependent stacks got lost somewhere at some point...

Reply:
But if user mis-indents (or makes typos) then output should say that those will be removed?

Reply:
Yes, but only during the first `upgrade`, and also not e.g. during the initial `stack install`